### ISSUE TITLE
Subsonic: load large playlist

### ISSUE THREAD

**Issue by jones1008** [ampache/ampache#1714]
# Environment

 * Ampache version: Ampache 3.8.6-release
 * Web server + version: Apache/2.4.25 (Debian)
 * Server operating system: Linux odroid-jessie 4.9.58+
 * Client operating system: iPhone iOS 11.3

# Settings

[http://paste.debian.net/1023582/](http://paste.debian.net/1023582/)

# Description

I am trying to load a large playlist (~ 20.000 tracks) on [my iOS subsonic client](https://itunes.apple.com/us/app/play-sub-music-streamer/id955329386?mt=8) and I always get the following ampache error within the ampache log after a minute trying to load the playlist:
```
2018-05-06 20:40:20 [ampache] (PHP) -> [Runtime Error] DOMDocument::loadXML(): invalid character in attribute value in Entity, line: 2 in file /var/www/ampache/lib/class/subsonic_api.class.php(191)
2018-05-06 20:40:20 [ampache] (PHP) -> [Runtime Error] DOMDocument::loadXML(): attributes construct error in Entity, line: 2 in file /var/www/ampache/lib/class/subsonic_api.class.php(191)
2018-05-06 20:40:20 [ampache] (PHP) -> [Runtime Error] DOMDocument::loadXML(): Couldn't find end of Start Tag entry line 2 in Entity, line: 2 in file /var/www/ampache/lib/class/subsonic_api.class.php(191)
2018-05-06 20:40:20 [ampache] (PHP) -> [Runtime Error] DOMDocument::loadXML(): PCDATA invalid Char value 11 in Entity, line: 2 in file /var/www/ampache/lib/class/subsonic_api.class.php(191)
```
The iOS client doesn't get anything back from the server. This is also recreateable with [this iOS app](https://itunes.apple.com/us/app/isub-music-streamer/id362920532)

At first I thought this would be a server memory problem and I increased the memory limit of the subsonic deamon as explained in this [subsonic forum thread](http://forum.subsonic.org/forum/viewtopic.php?f=6&t=3939&p=13801&hilit=increase+memory+limit#p13801) to `1024` but without any changes...

Now I think it is just a timeout problem within the `subsonic_api.class.php` but I don't know how to fix that...
Any ideas?

**Comment by wagnered** [comment]
@jones1008 
The above errors look like they are from the Ampache debug log, therefore I think you should be suspecting php parameters. It is php that is running the Ampache script:
1. max_execution_time = 60 (default)
2. max_input_time = 60 (default)
3. memory_limit = 128M (default)

These are found in the Linux php config file "php.ini".  I have increased 1 and 2 to 1440 to satisfy my scripting needs. But they should be kept as low a value as possible that gets the job done.

**Comment by jones1008** [comment]
thanks for the suggestion!
I changed the following php.ini configs to `120` (2 minutes):

1. `max_execution_time`
2. `max_input_time`
3. `default_socket_timeout`

I also changed the `memory_limit` to `1024`

After that I rebooted the system and tried to load the playlist again. 
So in theory if the script takes longer than 2 minutes it should throw the error... but it actually throws the error above after pretty much exactly 1 minute (I clocked it).

So is there maybe somewhere a local setting within the script that sets the timout to 60 seconds?

**Comment by wagnered** [comment]
Are you modifying the switches on the server?  I reduced my settings down to 120 seconds and restarted Apace and I still don't get the php errors in my ampache log as you do.

Anyway, I fired up my Subsonic server to try the playlist on it.  Subsonic informed me that 5 mb is the largest file it will accept. Mine was 8.9 MB with 23000 files.  So that is telling me that playlists are not meant to include the entire collection, but only a subset of maybe a particular genre, etc.

So I reduced the playlist down to just under 5 mb 13000+ titles in an m3u playlist and imported the playlist file into Ampache again.

I still don't get a php  error but I did notice that in the Apache access log indicated that accessing the playlist caused a "504" (internal) error.

So I am going to investigate this error, but please understand that because php is much slower that many other scripting languages, Ampache might never serve a playlist that large.(The subsonic server is written in Java.)

**Comment by jones1008** [comment]
I am modifying the php switches within the `php.ini` file on the server. The subsonic switch I am using is configured in the `/etc/default/subsonic` file (`SUBSONIC_ARGS="--max-memory=512"`)
My playlist file with ~ 20000 tracks is about 1.22MB big, so definitely under 5 MB

I don't get any 504 error from the apache access.log file...

Does the playlist successfully load on your client you tested it on?

**Comment by wagnered** [comment]
> Does the playlist successfully load on your client you tested it on?

From the Subsonic server, Ultrasonic  gets "Java heap space" error, possibly Ultrasonic problem, The Subsonic client displays the playlist with no problems, DSub also displays the playlist with no problems.

> My playlist file with ~ 20000 tracks is about 1.22MB big, so definitely under 5 MB

Interesting because my M3U playlist is 2.2 MB with 13000 titles.  Must be long titles :smile:

**Comment by jones1008** [comment]
> From the Subsonic server, Ultrasonic gets "Java heap space" error, possibly Ultrasonic problem, The Subsonic client displays the playlist with no problems, DSub also displays the playlist with no problems.

Mh... I tried a lot of iOS apps but with every app I am trying to open this huge playlist the server throws the same error.

> Interesting because my M3U playlist is 2.2 MB with 13000 titles. Must be long titles 😄

Yea, I use dynamic paths in my playlist file which makes it way smaller but the subsonic server does fetch the songs from the ampache database and not from the .m3u files... so why are we even discusssing about this?

So I downloaded the songlist of this huge playlist from the xml api and noticed that the .xml file is > 8MB with this:
`https://<host>/server/xml.server.php?action=playlist_songs&filter=<playlistid>&auth=<api_key>`
Maybe this xml file is too big for the subsonic server? 
How can I tell the subsonic server that it should accept files over 5MB?

I also had a look at the line [190 in subsonic_api.class.php](https://github.com/ampache/ampache/blob/develop/lib/class/subsonic_api.class.php#L190) where the error happens and then I read [this comment from the PHP loadXML documentation](http://php.net/manual/de/domdocument.loadxml.php#113676) which says that you should add `LIBXML_PARSEHUGE` to the `loadXML` function. 
So I replaced line 190 with `$dom->loadXML($xmlstr, LIBXML_PARSEHUGE);`
Sadly without success...

**Comment by wagnered** [comment]
> But it didn't output any new lines within the ampache log, so I don't know what's in the $xmlstr variable. 

Is your debug log set to level 5?
Are you still getting the "PCDATA invalid Char value 11 in Entity, line: 2" as described in your initial comment or is it not getting that far?  Because there might be some invalid metadata in one of  the songs.

> Do you think it crashes because the xml file size is above 5 MB? Or am I getting that wrong?

No, it just means there is more data to format and makes Ampache much slower in producing the output to send back to the client.. 

Just keep moving the debug_event statement back in the code until you get some output in the log. For example, move the the debug_event command  to before 181 and examine $xml. The URL you used doesn't indicate, but if $f is a json file request, the function might have returned output at line 182.

Then backtrack to addPlaylist() on the subsonic_xml_data.class.php script, unless you have created a smart playlist in Ampache. Then it would be addSmartPlaylist();

**Comment by wagnered** [comment]
I made  changes to increase the performance of the Subsonic API playlist access.  Using DSub for Android, and a playlist of 22000+ titles, it now takes only about 1.3 minutes to display the entire playlist on my phone.  The previous configuration took almost 10 minutes. I pushed the latest commit to the develop branch.

**Comment by jones1008** [comment]
Alright, thanks for your work!
My debug log level was set to 3 (*facepalm*)

When I added `debug_event('ampache', 'this is xmlstr: '.$xmlstr, 4);` to the [subsonic_api.class.php](https://github.com/ampache/ampache/blob/develop/lib/class/subsonic_api.class.php#L187) file (within the `else {}`) I got a huge xml string in the ampache log with all tracks and data in it...
But when I added `debug_event('ampache', 'this is xml: '.((string)$xml), 4);` to the [line 181](https://github.com/ampache/ampache/blob/develop/lib/class/subsonic_api.class.php#L181) I got an empty output like so: `2018-05-12 13:09:20 [jones] (ampache) -> this is xml:`
Should I maybe completely reinstall ampache and try it again?
Is it maybe because you are using an Android device and I am using an iOS device?

With the latest commit [e32bea7](https://github.com/ampache/ampache/commit/e32bea7bcd541fd9cae813b264fc324d676169c1) there is the following error in the ampache log when trying to get any song via the subsonic api:
`2018-05-12 12:25:18 [ampache] (PHP) -> [Error] Object of class Song could not be converted to string in file /var/www/ampache/lib/class/dba.class.php(85)`
It fully works when not using the latest commit.

**Comment by wagnered** [comment]
Sorry.  If you saw XML text at line 187, you should have then put the debug_event statement just before line 192.  Line 192 is the statement that pushes the data out to the web-server. 

> I got a huge xml string in the ampache log with all tracks and data in it...

If you see well formed XML at this point, That is,  the closing "</playlist>" tag at the end of the xml, it should be ok.

> With the latest commit e32bea7 there is the following error...

What about songs from the playlist?  I just tested the commit by successfully playing a song from the playlist.  Didn't check every menu item on the client. but I do see the error you refer to. Thanks.

**Comment by wagnered** [comment]
Seems to work completely on DSub for Android.  Excessive network retries accessing library on Subsonic for Android.  Difference is that DSub pages (default 20 user setting) display of the library.  Subsonic tries to get entire list at one time (NO timeout settings).

Changes pushed to develop branch.

**Comment by wagnered** [comment]
@jones1008 

Thanks for tracking down the problem.
 
Your suggestion will  probably be in the next commit pushed to the develop branch.

**Comment by lachlan-00** [comment]
Closing, This was fixed a while ago now in 3.8.8
